src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -713,7 +711,13 @@ public MetadataSnapshot(Map<String, StoreFileMetaData> metadata) {
                 if (maxVersion.onOrAfter(FIRST_LUCENE_CHECKSUM_VERSION)) {
                     checksumFromLuceneFile(directory, segmentsFile, builder, logger, maxVersion, true);
                 } else {
-                    builder.put(segmentsFile, new StoreFileMetaData(segmentsFile, directory.fileLength(segmentsFile), legacyChecksum, maxVersion, hashFile(directory, segmentsFile)));
+                    final BytesRefBuilder fileHash = new BytesRefBuilder();
+                    final long length;
+                    try (final IndexInput in = directory.openInput(segmentsFile, IOContext.READONCE)) {
+                        length = in.length();
+                        hashFile(fileHash, new InputStreamIndexInput(in, length), length);
```

Comment:
I think its awkward we wrap these files in InputStreamIndexInput here, and have the method take InputStream, when it could just take DataInput?

The one lone other usage of it, in a separate file (BlobStoreIndexShardRepository), could just wrap its InputStream with a o.a.l.util.InputStreamDataInput. and then the hashFile() would just be a simple readBytes() call into the byte array.

I know this isn't new in the patch, but this path adds more indexinput-wrapping since its not calling hashFile(String) anymore.